File: Java/src/main/java/com/gildedrose/ItemStrategy.java
package com.gildedrose;

public abstract class ItemStrategy {

    public abstract void update(Item item);

    public void decrementQuality(Item item){
        item.quality = item.quality - 1;
    };

    public void decrementSellin(Item item){
        item.sellIn = item.sellIn - 1;
    }

    public void incrementQuality(Item item) {
        item.quality = item.quality + 1;
    }
}

File: csharp/ItemStrategy.cs
using System;
namespace GildedRose
{
    public abstract class ItemStrategy
    {
        public void IncrementQuality(Item item)
        {
            if (item.Quality < 50)
            {
                item.Quality = item.Quality + 1;
            }
        }

        public virtual void DecrementQuality(Item item)
        {
            if (item.Quality > 0)
            {
                item.Quality = item.Quality - 1;

            }
        }

        public void DecrementSellin(Item item)
        {
            item.SellIn = item.SellIn - 1;
        }

        public abstract void Update(Item item);
    }
}
